File: benchmarks/compare.py
"""A/B delta report between two results files — the evidence format optimization PRs
cite. Warn-only by design: `compare` renders, it never gates (hosted-runner wall clock
is advisory; see docs/design/benchmarks.md)."""

from __future__ import annotations

from dataclasses import dataclass

from .results import Results, python_major_minor

# |Δ| must clear BOTH: 5% of base and a per-unit absolute noise floor. The floor is
# NOT one physical duration converted between units: macro (hyperfine, ms-scale)
# numbers carry hosted-runner process noise, so sub-2ms wiggle is meaningless there —
# while micro (pyperf, µs-scale) numbers are calibrated in-process loops whose values
# sit in the 1-600 µs range, where a 2000 µs floor would swallow every realistic
# regression (a 3x micro slowdown must land in "Notable", not "Within noise").
RELATIVE_THRESHOLD = 0.05
TIME_FLOOR_BY_UNIT = {"s": 0.002, "ms": 2.0, "us": 1.0}


@dataclass(frozen=True)
class Delta:
    metric: str
    unit: str
    base: float
    head: float

    @property
    def diff(self) -> float:
        return self.head - self.base

    @property
    def pct(self) -> float | None:
        if self.base == 0:
            return None
        return self.diff / self.base * 100

    @property
    def notable(self) -> bool:
        floor = TIME_FLOOR_BY_UNIT.get(self.unit, 0.0)
        if abs(self.diff) <= floor:
            return False
        if self.base == 0:
            return self.head != 0
        return abs(self.diff) > RELATIVE_THRESHOLD * abs(self.base)


@dataclass(frozen=True)
class Comparison:
    deltas: list[Delta]
    only_base: list[str]
    only_head: list[str]
    # Provenance fields on which the two sides differ ("profile: full vs pr", ...) —
    # non-empty means the delta table compares apples to oranges and the report says
    # so up top instead of rendering a clean-looking A/B.
    incomparable: list[str]

    @property
    def notable(self) -> list[Delta]:
        return [d for d in self.deltas if d.notable]


def compare(base: Results, head: Results) -> Comparison:
    """`pipeline.*` self-timings (suite durations, skip counts) are excluded: they
    measure the harness, not skit, and their wobble would intermittently pollute the
    "Notable" section of A/B evidence. Provenance mismatches (profile, platform,
    python) are detected here and rendered as a loud warning — a full-vs-pr or
    cross-platform diff must never look like a legitimate comparison."""
    incomparable = [
        f"{label}: {a} vs {b}"
        for label, a, b in (
            ("profile", base.meta.profile, head.meta.profile),
            ("platform", base.meta.host.platform_key, head.meta.host.platform_key),
            (
                "runner image",
                base.meta.host.ci_image_version,
                head.meta.host.ci_image_version,
            ),
            (
                "python",
                python_major_minor(base.meta.python),
                python_major_minor(head.meta.python),
            ),
            ("pyperf", base.meta.pyperf, head.meta.pyperf),
        )
        if a != b
    ]
    deltas: list[Delta] = []
    for metric_id in sorted(set(base.metrics) & set(head.metrics)):
        if metric_id.startswith("pipeline."):
            continue
        b, h = base.metrics[metric_id], head.metrics[metric_id]
        if b.unit != h.unit:
            incomparable.append(f"unit {metric_id}: {b.unit} vs {h.unit}")
            continue
        deltas.append(Delta(metric=metric_id, unit=h.unit, base=b.value, head=h.value))
    return Comparison(
        deltas=deltas,
        only_base=sorted(
            m for m in set(base.metrics) - set(head.metrics) if not m.startswith("pipeline.")
        ),
        only_head=sorted(
            m for m in set(head.metrics) - set(base.metrics) if not m.startswith("pipeline.")
        ),
        incomparable=incomparable,
    )


def render_markdown(base: Results, head: Results, comparison: Comparison) -> str:
    lines = [
        "## Benchmark comparison",
        "",
        f"Base: `{base.meta.git.commit[:12]}` ({base.meta.skit_version}) · "
        f"Head: `{head.meta.git.commit[:12]}` ({head.meta.skit_version}) · "
        f"profile {head.meta.profile} · {head.meta.host.platform_key}",
        "",
        "Warn-only: notable = |Δ| > max(5%, per-unit floor: 2 ms macro / 1 µs micro). "
        "Hosted-runner numbers are advisory.",
        "",
    ]
    if comparison.incomparable:
        lines += [
            "> [!WARNING]",
            "> **The sides are not directly comparable** — "
            + "; ".join(comparison.incomparable)
            + ". Deltas below mix apples and oranges.",
            "",
        ]
    notable = comparison.notable
    lines.append(f"### Notable ({len(notable)})" if notable else "### Notable (none)")
    if notable:
        lines += _table(notable)
    rest = [d for d in comparison.deltas if not d.notable]
    if rest:
        lines += ["", f"<details><summary>Within noise ({len(rest)})</summary>", ""]
        lines += _table(rest)
        lines += ["", "</details>"]
    for title, ids in (
        ("Only in base", comparison.only_base),
        ("Only in head", comparison.only_head),
    ):
        if ids:
            lines += ["", f"### {title}", ""]
            lines += [f"- `{metric_id}`" for metric_id in ids]
    return "\n".join(lines) + "\n"


def _table(deltas: list[Delta]) -> list[str]:
    rows = ["| Metric | Base | Head | Δ | Δ% |", "| --- | ---: | ---: | ---: | ---: |"]
    for d in deltas:
        pct = "—" if d.pct is None else f"{d.pct:+.1f}%"
        rows.append(
            f"| `{d.metric}` | {d.base:g} {d.unit} | {d.head:g} {d.unit} | {d.diff:+g} | {pct} |"
        )
    return rows

File: crates/skit-benchmarks/src/compare.rs
//! Warn-only A/B comparison reports.

use crate::{Results, budget::format_number};

const RELATIVE_THRESHOLD: f64 = 0.05;

/// One metric delta.
#[derive(Clone, Debug, PartialEq)]
pub struct Delta {
    /// Metric identifier.
    pub metric: String,
    /// Shared unit.
    pub unit: String,
    /// Base value.
    pub base: f64,
    /// Head value.
    pub head: f64,
}

impl Delta {
    /// Absolute signed change.
    #[must_use]
    pub fn difference(&self) -> f64 {
        self.head - self.base
    }

    /// Percentage change, absent when base is zero.
    #[must_use]
    pub fn percent(&self) -> Option<f64> {
        (self.base != 0.0).then(|| self.difference() / self.base * 100.0)
    }

    /// Whether the delta clears the per-unit noise policy.
    #[must_use]
    pub fn is_notable(&self) -> bool {
        if matches!(self.unit.as_str(), "count" | "bytes" | "bool") {
            return self.difference() != 0.0;
        }
        let floor = match self.unit.as_str() {
            "s" => 0.002,
            "ms" => 2.0,
            "us" => 1.0,
            _ => 0.0,
        };
        if self.difference().abs() <= floor {
            return false;
        }
        if self.base == 0.0 {
            return self.head != 0.0;
        }
        self.difference().abs() > RELATIVE_THRESHOLD * self.base.abs()
    }
}

/// Complete comparison.
#[derive(Clone, Debug, PartialEq)]
pub struct Comparison {
    /// Metrics available on both sides with matching units.
    pub deltas: Vec<Delta>,
    /// Metrics only in base.
    pub only_base: Vec<String>,
    /// Metrics only in head.
    pub only_head: Vec<String>,
    /// Provenance and unit mismatches.
    pub incomparable: Vec<String>,
}

impl Comparison {
    /// Deltas that clear the noise policy.
    #[must_use]
    pub fn notable(&self) -> Vec<&Delta> {
        self.deltas
            .iter()
            .filter(|delta| delta.is_notable())
            .collect()
    }
}

/// Compare two validated artifacts without gating them.
#[must_use]
pub fn compare(base: &Results, head: &Results) -> Comparison {
    let mut incomparable = Vec::new();
    mismatch(
        &mut incomparable,
        "profile",
        base.meta.profile.as_str(),
        head.meta.profile.as_str(),
    );
    mismatch(
        &mut incomparable,
        "platform",
        &base.meta.host.platform_key,
        &head.meta.host.platform_key,
    );
    mismatch_option(
        &mut incomparable,
        "runner image",
        base.meta.host.ci_image_version.as_deref(),
        head.meta.host.ci_image_version.as_deref(),
    );
    mismatch(
        &mut incomparable,
        "python",
        &crate::budget::python_major_minor(&base.meta.python),
        &crate::budget::python_major_minor(&head.meta.python),
    );
    mismatch(
        &mut incomparable,
        "pyperf",
        &base.meta.pyperf,
        &head.meta.pyperf,
    );

    let mut deltas = Vec::new();
    for (metric_id, base_metric) in &base.metrics {
        if metric_id.starts_with("pipeline.") {
            continue;
        }
        let Some(head_metric) = head.metrics.get(metric_id) else {
            continue;
        };
        if base_metric.unit != head_metric.unit {
            incomparable.push(format!(
                "unit {metric_id}: {} vs {}",
                base_metric.unit, head_metric.unit
            ));
            continue;
        }
        deltas.push(Delta {
            metric: metric_id.clone(),
            unit: head_metric.unit.clone(),
            base: base_metric.value,
            head: head_metric.value,
        });
    }
    let mut only_base = base
        .metrics
        .keys()
        .filter(|id| !id.starts_with("pipeline.") && !head.metrics.contains_key(*id))
        .cloned()
        .collect::<Vec<_>>();
    let mut only_head = head
        .metrics
        .keys()
        .filter(|id| !id.starts_with("pipeline.") && !base.metrics.contains_key(*id))
        .cloned()
        .collect::<Vec<_>>();
    only_base.sort();
    only_head.sort();
    Comparison {
        deltas,
        only_base,
        only_head,
        incomparable,
    }
}

fn mismatch(output: &mut Vec<String>, label: &str, base: &str, head: &str) {
    if base != head {
        output.push(format!("{label}: {base} vs {head}"));
    }
}

fn mismatch_option(output: &mut Vec<String>, label: &str, base: Option<&str>, head: Option<&str>) {
    if base != head {
        output.push(format!(
            "{label}: {} vs {}",
            base.unwrap_or("None"),
            head.unwrap_or("None")
        ));
    }
}

/// Render the evidence report.
#[must_use]
pub fn render_markdown(base: &Results, head: &Results, comparison: &Comparison) -> String {
    let mut lines = vec![
        "## Benchmark comparison".to_owned(),
        String::new(),
        format!(
            "Base: `{}` ({}) · Head: `{}` ({}) · profile {} · {}",
            prefix(&base.meta.git.commit, 12),
            base.meta.skit_version,
            prefix(&head.meta.git.commit, 12),
            head.meta.skit_version,
            head.meta.profile.as_str(),
            head.meta.host.platform_key
        ),
        String::new(),
        "Warn-only: notable = |Δ| > max(5%, per-unit floor: 2 ms macro / 1 µs micro); counts, byte sizes and booleans are exact — any change is notable. Hosted-runner numbers are advisory.".to_owned(),
        String::new(),
    ];
    if !comparison.incomparable.is_empty() {
        lines.extend([
            "> [!WARNING]".to_owned(),
            format!(
                "> **The sides are not directly comparable** — {}. Deltas below mix apples and oranges.",
                comparison.incomparable.join("; ")
            ),
            String::new(),
        ]);
    }
    let notable = comparison.notable();
    lines.push(if notable.is_empty() {
        "### Notable (none)".to_owned()
    } else {
        format!("### Notable ({})", notable.len())
    });
    if !notable.is_empty() {
        lines.extend(table(notable.into_iter()));
    }
    let noise = comparison
        .deltas
        .iter()
        .filter(|delta| !delta.is_notable())
        .collect::<Vec<_>>();
    if !noise.is_empty() {
        lines.extend([
            String::new(),
            format!("<details><summary>Within noise ({})</summary>", noise.len()),
            String::new(),
        ]);
        lines.extend(table(noise.into_iter()));
        lines.extend([String::new(), "</details>".to_owned()]);
    }
    append_only(&mut lines, "Only in base", &comparison.only_base);
    append_only(&mut lines, "Only in head", &comparison.only_head);
    append_skips(&mut lines, "base", base);
    append_skips(&mut lines, "head", head);
    lines.push(String::new());
    lines.join("\n")
}

fn table<'a>(deltas: impl Iterator<Item = &'a Delta>) -> Vec<String> {
    let mut lines = vec![
        "| Metric | Base | Head | Δ | Δ% |".to_owned(),
        "| --- | ---: | ---: | ---: | ---: |".to_owned(),
    ];
    for delta in deltas {
        let percent = delta
            .percent()
            .map_or_else(|| "—".to_owned(), |value| format!("{value:+.1}%"));
        lines.push(format!(
            "| `{}` | {} {} | {} {} | {} | {percent} |",
            delta.metric,
            format_number(delta.base),
            delta.unit,
            format_number(delta.head),
            delta.unit,
            format_signed(delta.difference())
        ));
    }
    lines
}

fn format_signed(value: f64) -> String {
    let sign = if value < 0.0 { '-' } else { '+' };
    format!("{sign}{}", format_number(value.abs()))
}

fn append_only(lines: &mut Vec<String>, title: &str, metrics: &[String]) {
    if metrics.is_empty() {
        return;
    }
    lines.extend([String::new(), format!("### {title}"), String::new()]);
    lines.extend(metrics.iter().map(|metric| format!("- `{metric}`")));
}

fn append_skips(lines: &mut Vec<String>, label: &str, results: &Results) {
    if results.skipped.is_empty() {
        return;
    }
    lines.extend([
        String::new(),
        format!("### Skipped in {label} ({})", results.skipped.len()),
        String::new(),
    ]);
    lines.extend(
        results
            .skipped
            .iter()
            .map(|skip| format!("- `{}/{}`: {}", skip.suite.as_str(), skip.case, skip.reason)),
    );
}

fn prefix(value: &str, length: usize) -> &str {
    value.get(..length).unwrap_or(value)
}
